### PR TITLE
Count lines of code in `template` and `style` of `.vue` files

### LINKED ISSUE #3251: Count lines of code in `template` and `style` of `.vue` files
reported in https://community.sonarsource.com/t/vue-templates-not-counted-in-total-lines-of-code/67959/5

### PR DESCRIPTION
**Pull request by yassin-kammoun-sonarsource**
Fixes #3251

### CODE REVIEW

File: eslint-bridge/src/linting/eslint/linter/visitors/metrics/ncloc.ts [hunk 1]
```diff
@@ -33,9 +34,68 @@ import { addLines } from './helpers';
...
+   * The template section parsed by `vue-eslint-parser` includes tokens for the whole `.vue` file.
+   * Everything that is not template-related is either raw text or whitespace. Although the style
+   * section is not parsed, its tokens are made available. Therefore, in addition to the tokens of
+   * the script section, we consider tokens from the template and style sections as well, provided
+   * that they don't denote whitespace or comments.
+   */
+  if (ast.templateBody) {
+    let withinStyle = false;
+    let withinComment = false;
+    for (const token of ast.templateBody.tokens) {
```
Comment (victor-diez-sonarsource) [inline eslint-bridge/src/linting/eslint/linter/visitors/metrics/ncloc.ts:58]: I feel that the different token types are treated in a very case by case scenario, are we sure we are covering all corner cases? I'm always afraid of these approaches

Comment (yassin-kammoun-sonarsource) [reply eslint-bridge/src/linting/eslint/linter/visitors/metrics/ncloc.ts:58]: The short answer: yes, I am sure.

Comment (yassin-kammoun-sonarsource) [reply eslint-bridge/src/linting/eslint/linter/visitors/metrics/ncloc.ts:58]: The long answer: yes, I am sure.

First, ESLint parsers distinguish tokens from comments and separate them in the abstract syntax tree as two different properties, namely `tokens` and `comments`. Furthermore, whitespace tokens are not included in any of those.

Now, `vue-eslint-parser` is expected to includes tokens from the `script` section as well as the `template` section:

- Tokens from the `script` section are made available from the root of the abstract syntax tree through the `tokens` property. These tokens follow ESLint convention and don't include whitespace tokens.
- Tokens from the `template` section HTML-specific tokens for the `template` section
- For some reason, tokens from the `template` section also include whitespace tokens as well as `HTMLRawText`-like tokens denoting tokens from the `script` and `style` section.

In the end, the list of scenarios is rather short:

1. We need to ignore all `HTMLRawText`-like tokens from the `script` section.
2. We need to ignore all `HTMLWhitespace` tokens from everywhere.
3. We need to ignore all comment-like tokens from the `style` section.

File: eslint-bridge/src/linting/eslint/linter/visitors/metrics/ncloc.ts [hunk 2]
```diff
@@ -33,9 +34,68 @@ import { addLines } from './helpers';
  */
 export function findNcloc(sourceCode: SourceCode): number[] {
   const lines: Set<number> = new Set();
-  const tokens = sourceCode.ast.tokens;
+  const tokens = [...sourceCode.ast.tokens, ...extractVuejsTokens(sourceCode)];
```
[the remark this answers is not in the document]
Comment (yassin-kammoun-sonarsource) [reply eslint-bridge/src/linting/eslint/linter/visitors/metrics/ncloc.ts]: 👍

File: eslint-bridge/src/linting/eslint/linter/visitors/metrics/ncloc.ts (hunk 1 above)
Comment (victor-diez-sonarsource) [reply eslint-bridge/src/linting/eslint/linter/visitors/metrics/ncloc.ts]: Thanks Yassin, much clearer!

**Review by victor-diez-sonarsource — APPROVED** [review]

**Pull request merged**